planner/core/rule_partition_processor.go

```diff
@@ -351,3 +362,445 @@ func (s *partitionProcessor) findByName(partitionNames []model.CIStr, partitionN
+	return len(or)
+}
+
+func (or partitionRangeOR) Less(i, j int) bool {
+	return or[i].start < or[j].start
+}
+
+func (or partitionRangeOR) Swap(i, j int) {
+	or[i], or[j] = or[j], or[i]
+}
+
+func (or partitionRangeOR) union(x partitionRangeOR) partitionRangeOR {
+	or = append(or, x...)
+	return or.simplify()
+}
+
+func (or partitionRangeOR) simplify() partitionRangeOR {
+	// Make the ranges order by start.
+	sort.Sort(or)
+	sorted := or
+
+	// Iterate the sorted ranges, merge the adjacent two when their range overlap.
+	// For example, [0, 1), [2, 7), [3, 5), ... => [0, 1), [2, 7) ...
+	res := sorted[:1]
+	for _, curr := range sorted[1:] {
+		last := &res[len(res)-1]
+		if curr.start > last.end {
+			res = append(res, curr)
+		} else {
+			// Merge two.
+			if curr.end > last.end {
+				last.end = curr.end
+			}
+		}
+	}
+	return res
+}
+
+func (or partitionRangeOR) intersection(x partitionRangeOR) partitionRangeOR {
+	if or.Len() == 1 {
+		return x.intersectionRange(or[0].start, or[0].end)
+	}
+	if x.Len() == 1 {
+		return or.intersectionRange(x[0].start, x[0].end)
+	}
+
+	// Rename to x, y where len(x) > len(y)
+	var y partitionRangeOR
+	if or.Len() > x.Len() {
+		x, y = or, x
+	} else {
+		y = or
+	}
+
+	// (a U b) M (c U d) => (x M c) U (x M d), x = (a U b)
+	res := make(partitionRangeOR, 0, len(y))
+	for _, r := range y {
+		// As intersectionRange modify the raw data, we have to make a copy.
+		tmp := make(partitionRangeOR, len(x))
+		copy(tmp, x)
+		tmp = tmp.intersectionRange(r.start, r.end)
+		res = append(res, tmp...)
+	}
+	return res.simplify()
+}
+
+// intersectionRange calculate the intersection of [start, end) and [newStart, newEnd)
+func intersectionRange(start, end, newStart, newEnd int) (int, int) {
+	var s, e int
+	if start > newStart {
+		s = start
+	} else {
+		s = newStart
+	}
+
+	if end < newEnd {
+		e = end
+	} else {
+		e = newEnd
+	}
+	return s, e
+}
+
+func (s *partitionProcessor) pruneRangePartition(ds *DataSource, pi *model.PartitionInfo) (LogicalPlan, error) {
+	lessThan, err := makeLessThanData(pi)
+	if err != nil {
+		return nil, err
+	}
+
+	col, fn, err := makePartitionByFnCol(ds, pi)
+	if err != nil {
+		return nil, err
+	}
+
+	result := fullRange(len(pi.Definitions))
+	if col != nil {
+		result = partitionRangeForCNFExpr(ds.ctx, ds.allConds, lessThan, col, fn, result)
+	}
+
+	return s.makeUnionAllChildren(ds, pi, result)
+}
+
+// makeLessThanData extracts the less than parts from 'partition p0 less than xx ... partitoin p1 less than ...'
+func makeLessThanData(pi *model.PartitionInfo) (lessThanData, error) {
+	var maxValue bool
+	lessThan := make([]int64, len(pi.Definitions))
+	for i := 0; i < len(pi.Definitions); i++ {
+		if strings.EqualFold(pi.Definitions[i].LessThan[0], "MAXVALUE") {
+			// Use a bool flag instead of math.MaxInt64 to avoid the corner cases.
+			maxValue = true
+		} else {
+			var err error
+			lessThan[i], err = strconv.ParseInt(pi.Definitions[i].LessThan[0], 10, 64)
```

Comment:
sql like `PARTITION p0 VALUES LESS THAN ( UNIX_TIMESTAMP('2008-01-01 00:00:00') )` will  not be supported?

Reply:
The `UNIX_TIMESTAMP('2008-01-01 00:00:00')` is already converted to an integer in `pi.Definitions`.

https://github.com/pingcap/tidb/blob/5778b50d0e0272aff0292dd96a323fa83f4e4aa1/ddl/partition.go#L430